src/kokkos_abstraction.hpp

```diff
@@ -118,65 +120,93 @@ static struct InnerLoopPatternTVR {
 static struct InnerLoopPatternSimdFor {
 } inner_loop_pattern_simdfor_tag;
 
+static struct ParallelForDispatch {
+} parallel_for_dispatch_tag;
+static struct ParallelReduceDispatch {
+} parallel_reduce_dispatch_tag;
+
+template <class... Args>
+struct DispatchType {
+  typedef typename std::conditional<sizeof...(Args) == 0, ParallelForDispatch,
+                                    ParallelReduceDispatch>::type type;
+};
+
+template <class... Args>
+inline void kokkos_dispatch(ParallelForDispatch, Args &&... args) {
+  Kokkos::parallel_for(std::forward<Args>(args)...);
+}
+template <class... Args>
+inline void kokkos_dispatch(ParallelReduceDispatch, Args &&... args) {
+  Kokkos::parallel_reduce(std::forward<Args>(args)...);
+}
+
```

Comment:
I'm torn about whether or not these dispatch types should be hidden inside an internal namespace. On the one hand, they're currently "implementation details." On the other hand, they could be convenient to expose.

Reply:
I introduced a namespace `dispatch_impl`.  what do you think @Yurlungur?

Reply:
Yep works for me.